### ISSUE TITLE
Sustainable public service from this code base is needed to make it useful.

### ISSUE THREAD

**Issue by mikedlr** [spamgourmet/spamgourmet#19]
At present the codebase here is divorced from the codebase which is running the actual spamgourmet service. This is unfortunate because my aim in setting up this project was not just to make the source code available for other people and provide a centre for maintaining it, but also to support the public spamgourmet service and as long as the code from here isn't in use for spamgourmet.com then we aren't really doing that properly.

The main reason for the separation is that there is some "secret" code which ensures that spamgourmet.com operates securely which isn't in a state that can be shared with this codebase.  It's unlikely (please correct this if wrong) that anyone who has access to the secret code will have time to tidy it up and secure it properly for public inclusion. This means that spamgourmet.com cannot just start running directly from the code base here. Unfortunately, due to the privacy commitments around spamgourmet it seems that it will be difficult to expand the list of people with access to the secret code. 

I'd like to ask for suggestions for how to get to the state where the code base here is able to run as the spamgourmet main code. 

One suggestion would be to start a new service (beta.spamgourmet.com? spamgourmetbeta.com? somethingelse.example.com?) running from this with a clear message that it isn't the production spamgourmet service but hopes to merge with it in future. 

Later, once the problems needed to run a public service were properly overcome, this code base would then be ready and could be used to take over service from spamgourmet.com.

**Comment by jgonggrijp** [comment]
> I'd like to ask for suggestions for how to get to the state where the code base here is able to run as the spamgourmet main code.

This feels like stating the obvious, but... I think the only way out is by removing the need for secrecy. However, it's impossible to give any concrete suggestions without knowing where the need for secrecy comes from.

Perhaps it would be possible for @josiah-hamilton to give a high-level explanation of the problem, without sharing the actual secret code, to a limited number of people? For example, just the current members of the spamgourmet organization, by email? In that way, at least thinking about the problem can be a team effort.

**Comment by mikedlr** [comment]
I agree that's starting from the existing system is likely to be the easiest way and that if @josiah-hamilton or one of the other old admins (who?) could support that would really help. However, I've seen that Josiah hasn't recently responded to messages so maybe he has other issues and won't be able to give time for now.

There is an alternative - start a clearly separate public service and from the beginning make it clear that this is a "beta" service for the real one which has lower security and privacy protection as well as no promises on reliability but will get new features first. From running that service and identifying the problems that come up we can learn not only lessons that we might learn from the private code, but probably also ones that are completely forgotten but protected against in ways that aren't obvious. 

Long run, once the service was clearly fixed, the code base could then be put onto the main spamgourmet server or something similar.  For this to work some admin support would be needed. It could happily run from a docker image in AWS or something else similarly cheap.

**Comment by vasile-gh** [comment]
I also agree. I would, in addition, like to explore making this a one-click
self-deployed install, because this is the only realistic way to avoid
being blacklisted as a disposable email.

Finally, exploring enabling ARC could also be something to look into.

**Comment by mikedlr** [comment]
I agree that if there's to be a future of this codebase then it needs to be packaged up for easy use.  If nothing else that's critical for allowing people who want to do testing and small amounts of development.  There are other features that I think are important at this stage - I think that use spamgourmet to run an email address hosted elsewhere shadowing addresses on that system (see my recent posts in the spamgourmet forum) would be transformational. @vasile-gh I think it would be good to open a new ticket about that and keep discussions about future directions separate. 

At this point  though something has to change.  I still really haven't found an adequate replacement for spamgourmet, so I really want it to work, however if there's nobody fixing problems then it will fail at some point and be irrecoverable.  

* should we revisit getting access to the private code?  I don't think the old admins currently have bandwidth to deal with things.  Maybe if we had a way of discussing with them it would help?
* if the admins can't provide support, the key thing that can't be transferred is the history of mail accounts - could we have an organisation which could inherit domains but without any private data? - allowing people to export that themselves would be a solution to maintaining a proper service
* is the proposal here - setting up a public service a step towards that?
* is there enough commitment available to support this?  I don't see much activity on the forums.

**Comment by vasile-gh** [comment]
I would simply start anew as "beta", without the private code.

I realise we risk a complete fork but absent support from spamgourmet, what can we do? And waiting is less and less of a good option.

**Comment by mikedlr** [comment]
Josiah has responded in #20 - I'm hoping he will give us some comment here.  I'm not sure if that changes anything, however I think that if there's anything that can be done to ensure that the main Spamgourmet service remains stable that should be the highest priority.

The main thing that I can see is the need to have more than one admin (preferably minimum 3) on the main service.  If Josiah has to be unavailable then there needs to be someone to take over.  If I understand correctly there isn't such a person right now? If we can find a way to provide that backup then all would be great. If there isn't an acceptable person available then going through the beta service and then maybe being ready to take responsibility for the domains without getting the email data would be a reasonable end game for the service???

**Comment by vasile-gh** [comment]
We should define the duties associated to this role before we seek volunteers. Josiah is basically doing everything but I am sure this is not needed for most admin duties.

**Comment by jgonggrijp** [comment]
I feel it's getting a bit muddy here... are we talking about the sustainability of the service or of the codebase? I honestly feel that only the latter is our concern. If the code in this repo does not fully represent the service and is in fact not entirely secure, than that is the problem I feel we should be solving. @josiah-hamilton has taken responsibility for the service and as long as he is publicly responding and not explicitly resigning, I feel we should trust him to do what is needed to keep the service up. There is definitely something to say for redundancy, but it is his choice to make that call.

@josiah-hamilton Please do reply here if you can find the time.

**Comment by mikedlr** [comment]
@jgonggrijp I think we largely agree - the primary priority on this project has to be the spamgourmet code - however this is an open source project and so it should be a meeting place of people with different interests working together, for example:

1. those that want to run their own service using the code 
2. those that want to support the existing spamgourmet service
3. those that want to have a long running public spamgourmet service independent of what happens to the original one

Group 2. and 3. are not identical in interests.  Group 2 is prioritizing the protection of the old spamgourmet data which they are responsible for. Group 3 may not need that data.

I think it's completely reasonable for group 1 and 3 to cooperate to set up a new service - initially as a test - potentially in future as a proper production service.  Given the nature of FOSS software, it would be nice if they took into account any wishes or requests from group 2 but that is entirely a matter of good will. What they *must* not do, however, is represent themselves as the "Spamgourmet" service.  "Spamlover, using the Spamgourmet code" would be fine.

**Comment by eegeeZA** [comment]
This repo started out as a snapshot which what I thought was the full source code. Currently we know that it is an incomplete snapshot because certain code was concealed to prevent abuse. So, although spamgourmet is opensource, it is only partially so, since there were also minor adjustments post the snapshot. From my understanding the current code is good enough to use for _personal_ hosting without the (D)DOS security code, and that is why I think Josh felt comfortable sharing the code for others to use.

While it would be nice for this to become the official repository, I believe that would require buy in from Josiah and the admins. If they are comfortable keeping their own repository, this could become place for them to create occasional pull request with sanitised code for public use. If they wanted to fully adopt this as their public repo, they could then engage with the opensource community for enhancements and fixes.

In the end, I do agree that it would be great if this became the repo for spamgourmet without any modifications compared to the PROD release. But I know that it is mostly out of our control. If there is a need to run the next spamgourmet v2 (beta) off this code, I would suggest creating a fork to include all the modifications and security that would need to be added.
